rust-driver/src/types.rs

```diff
@@ -0,0 +1,451 @@
+        // So we need to convert it to little endian, use `to_be()`
+        Self::new(val.to_be())
+    }
+}
+
+impl From<u32> for Key {
+    fn from(key: u32) -> Self {
+        Self::new(key)
+    }
+}
+
+/// Packet Sequence Number
+pub type Psn = ThreeBytesStruct;
+
+/// Queue Pair Number
+pub type Qpn = ThreeBytesStruct;
+
+/// In RDMA spec, some structs are defined as 24 bits.
+/// For example : `PSN`, `QPN` etc.
+///
+/// This struct is used to represent these 24 bits.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Default)]
+pub struct ThreeBytesStruct(u32);
+
+impl ThreeBytesStruct {
+    const WIDTH: usize = 24;
+    const MASK: u32 = u32::MAX >> (32 - Self::WIDTH);
+    const BORDER: u32 = Self::MASK + 1;
+    #[cfg(test)]
+    pub(crate) const MAX_VALUE: u32 = Self::MASK;
+
+    /// Create a new `ThreeBytesStruct` with the given value.
+    ///
+    /// If the value is greater than 24 bits, the higher bits will be ignored.
+    #[must_use]
+    pub fn new(key: u32) -> Self {
+        Self(key & Self::MASK)
+    }
+
+    /// Get the value of `ThreeBytesStruct`.
+    #[must_use]
+    pub fn get(&self) -> u32 {
+        self.0
+    }
+
+    /// Convert the value of `ThreeBytesStruct` to big endian.
+    #[must_use]
+    pub(crate) fn into_be(self) -> u32 {
+        // In little endian machine, to_le_bytes() is a no-op. Just get the layout.
+        let key = self.0.to_le_bytes();
+        // Then we reoder the bytes to big endian
+        // Note that the last byte is exceed the 24 bits, any value in it will be ignored
+        u32::from_le_bytes([key[2], key[1], key[0], 0])
+    }
+
+    /// Convert a big endian value to `ThreeBytesStruct`.
+    #[must_use]
+    pub(crate) fn from_be(val: u32) -> Self {
+        // get the layout.
+        let key = val.to_le_bytes();
+        // from_le_bytes is also a no-op in little endian machine.
+        // We just use it to convert from [u8;4] to `u32`.
+        Self::new(u32::from_le_bytes([key[2], key[1], key[0], 0]))
+    }
+
+    /// wrapping add the current value with rhs
+    #[must_use]
+    #[allow(clippy::arithmetic_side_effects)]
+    pub(crate) fn wrapping_add(self, rhs: Self) -> Self {
+        // since (a+b) mod p  = (a + (b mod p)) mod p, we don't have to let rhs= rhs%p here
+        Self((self.0 + rhs.0) % Self::BORDER)
+    }
+
+    /// wrapping sub the current value with rhs
+    #[must_use]
+    #[allow(clippy::arithmetic_side_effects)]
+    pub(crate) fn wrapping_sub(self, rhs: Self) -> Self {
+        if self.0 >= rhs.0 {
+            Self(self.0 - rhs.0)
+        } else {
+            Self(Self::BORDER - rhs.0 + self.0)
+        }
+    }
+
+    /// The absolute difference between two PSN
+    /// We assume that the bigger PSN should not exceed the
+    /// smaller PSN by more than 2^23(that half of the range)
+    #[must_use]
+    #[allow(clippy::arithmetic_side_effects)]
+    pub(crate) fn wrapping_abs(self, rhs: Psn) -> u32 {
```

Comment:
By read the comment, I think the ThreeBytesStruct is used to define Psn. So I think the impl block of ThreeBytesStruct should not have Psn.

Reply:
Yes, I think so. Although the `QPN` is 3 bytes, we should not provide arithmetic operations for it.